cartoframes/utils.py

```diff
@@ -15,3 +15,33 @@ def cssify(css_dict):
                                                      field_value=field_value)
         css += '} '
     return css
+
+def norm_colname(colname):
```

Comment:
@juanignaciosl, does this match what you would expect for column name conversion via the Import API? There are tests below too: https://github.com/CartoDB/cartoframes/pull/190/files#diff-85c7da7d97a696e8017da5742a876475R90

Reply:
Your test cases seem ok. You might want to review the following methods, which are the ones that we use to normalize: [sanitize_column_name](https://github.com/CartoDB/cartodb/blob/8e3b44622a89c3dd8b49fa8235a08984e7481da5/config/initializers/string.rb#L188-L198), [sanitize](https://github.com/CartoDB/cartodb/blob/8e3b44622a89c3dd8b49fa8235a08984e7481da5/config/initializers/string.rb#L127-L132), and [normalize](https://github.com/CartoDB/cartodb/blob/8e3b44622a89c3dd8b49fa8235a08984e7481da5/config/initializers/string.rb#L8). As you see, you might be skipping some special character transformations.

Logic is not equivalent for some corner cases. Example:

```ruby
2.2.3 :010 > sanitize('wadu<s<>a')
 => "wadua"
2.2.3 :012 >   sanitize('a&b.c+d?e;f')
 => "a_f"
```

vs

```python
>>> norm_colname('wadu<s<>a')
'wadu_s_a'
>>> norm_colname('a&b.c+d?e;f')
'a_b_c_d_e_f'
```

Nevertheless, it looks like your method generates valid column names. Is that enough for you, or are you looking for a perfect match?

PS: during backend retreat javitonino did a nice demo of cartoframes, I hope that we'll be able to add [overwrite collision strategy](https://github.com/CartoDB/cartodb/issues/11990) soon :_). Sorry about the delay, we're swamped :(

Reply:
The goal is to match what the Import API does. I opened this ticket (https://github.com/CartoDB/cartodb/issues/12331) to see if it would be feasible to get resultant schema. I understand your reasoning now.

I'm now realizing that I could change the column names before uploading by using my `norm_colname` method (which seems good enough). If I do that, the sanitizing methods you linked to will probably not need to do any extra processing and the column names probably won't be changed. Does that sound like a reasonable approach?

Reply:
> change the column names before uploading

Yes, that approach sounds really good to me. Maybe you'll have to do some adjustments later, but it will be much easier than trying to replicate importer (which might change).

Did you consider doing a request to SQL API as I suggested at CartoDB/cartodb/issues/12331? That will return the schema.

Reply:
Cool, thanks for the input!

I want to send the data without type guessing to avoid any unintended losses due to conversions of data (e.g., US Zip codes should be text and sometimes start with a 0 like `01391`, but are often interpreted as `numeric` so the leading zero is lost), and then enforce the schema afterwards. If I could tell the Import API the schema before import, that'd be cool ;). The process cartoframes uses is:
1. Send DataFrame (output to a CSV) through Import API with no type guessing (all columns will be text)
2. Once fully uploaded, run a large `ALTER TABLE "foo" ALTER COLUMN "unnamed_0" TYPE numeric USING NULLIF("unnamed_0", '')::numeric, ...`

The problem I see with getting the schema from the SQL API is that I would still have to make guesses about which columns got renamed and make that mapping. To me it feels safer to change the column names beforehand (assuming the normalizing function is good enough!).

Reply:
Ok, let's normalize beforehand then.

I'll do my best to make the overwrite strategy available ASAP :-)

Reply:
Cool, thanks for all the input @juanignaciosl, it's been really helpful!!
And looking forward to the overwrite feature :)